Shopper → Cart → Can proceed to checkout

tarhi-saad opened the pull request:
Fixes #5763

### Notes

This PR adds an e2e test to verify that a shopper can proceed to checkout from the Cart block by clicking on the `Proceed to Checkout` button.

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

### Manual Testing

1. Check out this PR.
2. Start Docker Desktop.
3. Run `npm run wp-env destroy && npm run wp-env start` to start a fresh e2e environment.
4. Run `npm run test:e2e -- tests/e2e/specs/shopper/cart-proceed-to-checkout.test.js` to run only this test case.
5. Run `npm run test:e2e` to run the whole test suite.

Commit:
Shopper → Cart → Can proceed to checkout

alexflorisca commented:
Hey @tarhi-saad, nice work in this PR! There is a lot of overlap with a couple of open PRs however so best to hold off for a minute while we merged that in. For example, I've already created the `goToCartBlock` function in [this PR](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5900) and @ralucaStan has made it more efficient in [this PR](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5938). 

I'm also wondering if we need this to be a separate test, as it is implicitly tested in #5900 (and probably will be in others that follow a similar format). @nielslange @ralucaStan what do you think?

Commit:
Merge branch 'trunk' into add/cart-proceed-to-checkout-e2e-test

nielslange commented:
> I'm also wondering if we need this to be a separate test, as it is implicitly tested in #5900 (and probably will be in others that follow a similar format). @nielslange @ralucaStan what do you think?

Great question, @alexflorisca. Initially, I had the same thought in mind. However, #5900 is using `await shopper.goToCheckoutBlock();` to go to the checkout block via URL. In comparison, this PR does not navigate to the  checkout block via URL, but by clicking the `Proceed to checkout` button on the cart page. Therefore, I'd say that this PR is still relevant. 

What are your thoughts, @ralucaStan?

nielslange reviewed (commented):
Since we merged #5938, could you rebase your PR against that PR which already has implementations both for `goToCheckoutBlock()` and `goToCartBlock()`?

Commit:
Refactor the code for E2E testing

Introduce some discussed solution to structure and optimize the code

tarhi-saad committed:
Merge branch 'trunk' into add/cart-proceed-to-checkout-e2e-test

Commit:
Resolve some conflicts with the trunk branch

tarhi-saad commented:
Hi @nielslange! I refactored the code to include some discussed code structure and optimization solutions. (e.g., using the `block` object inside `shopper`, adding constant of the simple product name, the use of `addToCartFromShopPage` method, etc.).

I also remove the method `shopper.emptyCart()` overriding the exiting one since we are trying to avoid doing that to prevent any confusion. And, instead, we can use our own "shopper.block.emptyCart()" method. I made sure to refactor the code accordingly.

Commit:
Remove method overriding and refactor code

We are trying to avoid overriding exiting E2E Woo's utility functions

Instead we can use our own "shopper.block"

Commit:
Use a consistent constant naming convention

ralucaStan commented on tests/utils/shopper.js:
```diff
@@ -61,37 +61,43 @@ export const shopper = {
-				products = await page.$$( '.remove' );
+		/**
+		 * For some reason "wcShopper.emptyCart" sometimes result in an error, but using the same
+		 * implementation here fixes the problem.
```
Looks like @dinhtungdu also fixed this, and created an issue to revert changes once the upstream fix is made. https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5966

tarhi-saad replied:
@ralucaStan. I copy-pasted the same answer from slack here, as it answers your question:

I agree with that. I have indeed rebased and I noticed @dinhtungdu's changes regarding the `emptyCart` fix. So, in the same PR, I refactored the code by removing the one overriding the Woo method & only keeping the newly added `shopper.block.emptyCart`, see this commit [Remove method overriding and refactor code](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5901/commits/7f9e4a7e631aba6a9669d7083657d6dda00cac8e).

tarhi-saad committed:
Merge branch 'trunk' into add/cart-proceed-to-checkout-e2e-test

tarhi-saad commented:
Hi @nielslange & @alexflorisca! It would be great to review and merge this PR since it introduces various ideas we discussed in the E2E Testing P2 (e.g., using the `block` object inside `shopper`, adding constant of the simple product name, the use of `addToCartFromShopPage` method, etc.). See this [comment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5901#issuecomment-1055804590) for more details!

alexflorisca commented on tests/e2e/specs/shopper/cart-proceed-to-checkout.test.js:
```diff
@@ -0,0 +1,40 @@
+	test.only( `skipping ${ block.name } tests`, () => {} );
+
+describe( 'Shopper → Cart → Can proceed to checkout', () => {
+	beforeAll( async () => {
```
instead emptying the cart `beforeAll` and `afterAll`, what do you think about emptying the cart in `beforeEach` so that if in the future more tests are added to this file, we don't need to change this?

tarhi-saad replied:
I agree with the `beforeEach`. I'll apply this in my next PR as well! 🙌

alexflorisca commented on tests/e2e/specs/shopper/cart-proceed-to-checkout.test.js:
```diff
@@ -0,0 +1,40 @@
+		await shopper.block.goToCart();
+
+		// Click on "Proceed to Checkout" button
+		await Promise.all( [
```
You could extract this into a `shopper.block.proceedToCheckout()` function as I think I wrote something similar for my test, and will likely be used by other tests

tarhi-saad replied:
This is a great idea, @alexflorisca! Thank you, I'll apply this on the next PR since I did go ahead and merge this one!

alexflorisca reviewed (approved):
Looks good, a couple of minor comments, but they are not a blocker for merging this PR . Nice work, a great framework from which to build the other test cases :)
